### ISSUE TITLE
incompatible types: `Optional<@Nullable SomeType>` cannot be converted to `Optional<SomeType>`

### ISSUE THREAD

**Issue by throup** [uber/NullAway#1594]
## Description

NullAway 0.13.6 is inferring `Optional`s of `@Nullable` types, whereas NullAway 0.13.4 had correctly derived them as non-null.

We have encountered this in a large Spring Boot / gradle project, and upgrading NullAway introduces hundreds of false positives of this kind. We are using JSpecify annotations and running in JSpecify mode.

The examples we have seen, so far, are generally from Stream terminal operations.

## Sample warning message
```
warning: [NullAway] incompatible types: Optional<@Nullable Double> cannot be converted to Optional<Double>
          .reduce(Double::sum);
                 ^
      (see http://t.uber.com/nullaway )
  1 error
  1 warning
```

## Minimal repro

```java
package example.nullaway;

import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;
import org.jspecify.annotations.Nullable;

public class NullAwayExample {
  // Both NullAway 0.13.4 and 0.13.6 accept this method with no violations.
  public static Optional<Double> sumNumbersAsDoubles(@Nullable Number... numbers) {
    return Arrays.stream(numbers)
        .filter(Objects::nonNull)
        .map(Number::doubleValue)
        .reduce(Double::sum);
  }

  // NullAway 0.13.4 accepts this method with no violations.
  //
  // NullAway 0.13.6 does not accept this method, and reports:
  // ```
  // warning: [NullAway] incompatible types: Optional<@Nullable Double> cannot be converted to Optional<Double>
  //          .reduce(Double::sum);
  //                 ^
  //      (see http://t.uber.com/nullaway )
  //  1 error
  //  1 warning
  // ```
  public static Optional<Double> sumDoubles(@Nullable Double... doubles) {
    return Arrays.stream(doubles)
        .filter(Objects::nonNull)
        .reduce(Double::sum);
  }

  // Both NullAway 0.13.4 and 0.13.6 accept this method with no violations.
  public static Optional<Double> sumDoublesWithExplicitMap(@Nullable Double... doubles) {
    return Arrays.stream(doubles)
        .filter(Objects::nonNull)
        .map(Objects::requireNonNull)
        .reduce(Double::sum);
  }
}
```

**Comment by lbenedetto** [comment]
Also happens for very simple cases like this one:
```java
public record Repro(@Nullable String name) {
  public Optional<String> getNameOpt() {
    return Optional.ofNullable(name)
        .filter(it -> true);
  }
}
```

**Comment by msridhar** [comment]
Thank you for all the examples.  I _think_ this is related to #1578, but also to library models for `Optional`.  Will prioritize fixing this one.

**Comment by lbenedetto** [comment]
You only fixed Streams, but this issue was also about Optional. I commented a reproducer.
I don't think this issue should be marked fixed until Optional filtering is also fixed.

**Comment by msridhar** [comment]
@lbenedetto I think your issue is a different bug than what was fixed here.  Let's track in #1611.
